applications/GeoMechanicsApplication/tests/cpp_tests/test_nodal_extrapolator.cpp

```diff
@@ -0,0 +1,128 @@
+    auto integration_points = geometry.IntegrationPoints(integration_method);
+    auto extrapolation_matrix =
+        nodal_extrapolator.CalculateElementExtrapolationMatrix(geometry, integration_method);
+
+    // clang-format off
+    Matrix expected_extrapolation_matrix = ZeroMatrix(3, 3);
+    expected_extrapolation_matrix <<= 5.0/3.0, -1.0/3.0, -1.0/3.0,
+                                     -1.0/3.0,  5.0/3.0, -1.0/3.0,
+                                     -1.0/3.0, -1.0/3.0,  5.0/3.0;
+    // clang-format on
+
+    KRATOS_EXPECT_MATRIX_NEAR(extrapolation_matrix, expected_extrapolation_matrix, 1e-6)
+}
+
+KRATOS_TEST_CASE_IN_SUITE(NodalExtrapolator_GivesCorrectExtrapolationMatrix_For2D6NTriangle, KratosGeoMechanicsFastSuite)
+{
+    Kratos::Triangle2D6<Node> geometry(
+        Kratos::make_intrusive<Node>(1, 0.0, 0.0, 0.0), Kratos::make_intrusive<Node>(2, 1.0, 0.0, 0.0),
+        Kratos::make_intrusive<Node>(3, 0.0, 1.0, 0.0), Kratos::make_intrusive<Node>(4, 0.5, 0.0, 0.0),
+        Kratos::make_intrusive<Node>(5, 0.5, 0.5, 0.0), Kratos::make_intrusive<Node>(6, 0.0, 0.5, 0.0));
+
+    NodalExtrapolator nodal_extrapolator;
+
+    auto integration_method = GeometryData::IntegrationMethod::GI_GAUSS_2;
+    auto integration_points = geometry.IntegrationPoints(integration_method);
+    auto extrapolation_matrix =
+        nodal_extrapolator.CalculateElementExtrapolationMatrix(geometry, integration_method);
+
+    // clang-format off
+    Matrix expected_extrapolation_matrix = ZeroMatrix(6, 3);
+    expected_extrapolation_matrix <<= 5.0/3.0, -1.0/3.0, -1.0/3.0,
+                                     -1.0/3.0,  5.0/3.0, -1.0/3.0,
+                                     -1.0/3.0, -1.0/3.0,  5.0/3.0,
+                                      2.0/3.0,  2.0/3.0, -1.0/3.0,
+                                     -1.0/3.0,  2.0/3.0,  2.0/3.0,
+                                      2.0/3.0, -1.0/3.0,  2.0/3.0;
+    // clang-format on
+
+    KRATOS_EXPECT_MATRIX_NEAR(extrapolation_matrix, expected_extrapolation_matrix, 1e-6)
+}
+
+KRATOS_TEST_CASE_IN_SUITE(NodalExtrapolator_GivesCorrectExtrapolationMatrix_For2D4NQuadrilateral, KratosGeoMechanicsFastSuite)
+{
+    Kratos::Quadrilateral2D4<Node> geometry(
+        Kratos::make_intrusive<Node>(1, 0.0, 0.0, 0.0), Kratos::make_intrusive<Node>(2, 1.0, 0.0, 0.0),
+        Kratos::make_intrusive<Node>(3, 1.0, 1.0, 0.0), Kratos::make_intrusive<Node>(4, 0.0, 1.0, 0.0));
+
+    NodalExtrapolator nodal_extrapolator;
+
+    auto integration_method = GeometryData::IntegrationMethod::GI_GAUSS_2;
+    auto integration_points = geometry.IntegrationPoints(integration_method);
+    auto extrapolation_matrix =
+        nodal_extrapolator.CalculateElementExtrapolationMatrix(geometry, integration_method);
+
+    // clang-format off
+    Matrix expected_extrapolation_matrix = ZeroMatrix(4, 4);
+    expected_extrapolation_matrix <<= 1.8660254037844386, -0.5,  0.13397459621556132, -0.5,
```

Comment:
formatting this differently may reveal the structure of the expected matrix more clearly. That holds further down this file too.